tracer/src/Datadog.Trace/Configuration/ExtraServicesProvider.cs

```diff
@@ -0,0 +1,47 @@
+﻿// <copyright file="ExtraServicesProvider.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.Configuration;
+
+internal class ExtraServicesProvider : IExtraServicesProvider
+{
+    private const int MaxExtraServices = 64;
+    private const string FakeValue = null;
+
+    // no concurrent hash set, so use a dictionary with empty values
+    private readonly ConcurrentDictionary<string, string> _extraServices = new();
+
+    public void AddService(string serviceName)
+    {
+        // Several threads entering simultaneously can cause MaxExtraService to be exceeded.
+        // As long as the list doesn't grow much beyond MaxExtraService we don't care.
+        if (_extraServices.Count < MaxExtraServices)
+        {
+            _extraServices.AddOrUpdate(serviceName, FakeValue, (string _, string _) => null!);
+        }
```

Comment:
The suggested snippet doesn't work since it increments the counter every time, not just when a service name is successfully added.

I modified the code to store the counter externally, but it requires a call to `ContainsKey`. From a quick look at the code it seems `Count` takes locks, but `ContainsKey` doesn't so perhaps is faster?

Reply:
see 14f828b6ca89a2f776967e35a647b6493ec40c97 for the change.

Reply:
> The suggested snippet doesn't work since it increments the counter every time,

Ah good point. I think you can still avoid the `ContainsKey`, given that you don't care about updates, I think you could just use `TryAdd()`? 

e.g. 

```csharp
private int _serviceCount = 0;

if (_serviceCount < MaxExtraServices)
{
    var wasAdded = _extraServices.TryAdd(serviceName, FakeValue);
    if (wasAdded)
    {
        Interlocked.Increment(ref _serviceCount);
    }
}

```

Reply:
If we assume that the service names within a service is fairly small and static, then most common cases is:

* current solution: Only `ContainsKey` is called because key is already present
* suggested solution: `TryAdd` is called but fails to add key

So the better solution is which ever is faster in that case, I had no idea which that is, so I wrote a benchmark and it seems `ContainsKey` is faster (on net7.0 at least):

```
  [Host]     : .NET 7.0.9 (7.0.923.32018), X64 RyuJIT AVX2
  Job-KWLFIE : .NET 7.0.9 (7.0.923.32018), X64 RyuJIT AVX2

Runtime=.NET 7.0  Toolchain=net7.0

|      Method |     Mean |    Error |   StdDev |   Median | Allocated |
|------------ |---------:|---------:|---------:|---------:|----------:|
|      TryAdd | 60.75 ns | 1.784 ns | 5.147 ns | 59.72 ns |         - |
| ContainsKey | 28.21 ns | 0.745 ns | 2.185 ns | 27.41 ns |         - |
```

Benchmark code (not committed):
```
    public class ConcurrentDictBenchmark
    {

        private const string TestServiceName = "service-name";
        private static ConcurrentDictionary<string, string> _concurrentDictionary;

        static ConcurrentDictBenchmark()
        {
            _concurrentDictionary = new ConcurrentDictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            _concurrentDictionary.TryAdd(TestServiceName, null);
        }

        [Benchmark]
        public void TryAdd() => _concurrentDictionary.TryAdd(TestServiceName, null);

        [Benchmark]
        public void ContainsKey() => _concurrentDictionary.ContainsKey(TestServiceName);
    }
```

Reply:
Cool, in which case I agree with current 👍 

Purely out of interest, can you try on netfx too if that's easy (no worries if not)? 😄

Reply:
netfx is slower but the `ContainsKey` is still faster:

```
|      Method |              Runtime |     Mean |    Error |   StdDev | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------ |--------------------- |---------:|---------:|---------:|------:|--------:|----------:|------------:|
|      TryAdd |             .NET 7.0 | 43.71 ns | 0.839 ns | 0.785 ns |  0.70 |    0.01 |         - |          NA |
|      TryAdd | .NET Framework 4.7.2 | 62.20 ns | 1.012 ns | 0.947 ns |  1.00 |    0.00 |         - |          NA |
|             |                      |          |          |          |       |         |           |             |
| ContainsKey |             .NET 7.0 | 21.11 ns | 0.441 ns | 0.433 ns |  0.53 |    0.02 |         - |          NA |
| ContainsKey | .NET Framework 4.7.2 | 39.76 ns | 0.744 ns | 0.913 ns |  1.00 |    0.00 |         - |          NA |```